```
In file included from ../librz/include/rz_vector.h:5,
                 from ../librz/include/rz_util/rz_event.h:12,
                 from ../librz/include/rz_util.h:28,
                 from ../librz/include/rz_bin.h:4,
                 from ../librz/bin/bobj.c:6:
../librz/bin/bobj.c: In function ‘rz_bin_object_p2v_all’:
../librz/include/rz_util/rz_assert.h:112:11: error: returning ‘long long unsigned int’ from a function with return type ‘RzVector *’ {aka ‘struct rz_vector_t *’} makes pointer from integer without a cast [-Werror=int-conversion]
  112 |    return (val); \
      |           ^
../librz/bin/bobj.c:998:2: note: in expansion of macro ‘rz_return_val_if_fail’
  998 |  rz_return_val_if_fail(obj, UT64_MAX);
      |  ^~~~~~~~~~~~~~~~~~~~~
```

--- a/librz/bin/bin.c
+++ b/librz/bin/bin.c
@@ -649,6 +649,14 @@ RZ_API RzList *rz_bin_get_sections(RzBin *bin) {
 	return o ? (RzList *)rz_bin_object_get_sections_all(o) : NULL;
 }
 
+/**
+ * \brief Find the binary section at offset \p off.
+ *
+ * \param o Reference to the \p RzBinObject instance
+ * \param off Address to search
+ * \param va When 0 the offset \p off is considered a physical address, otherwise a virtual address
+ * \return Pointer to a \p RzBinSection containing the address
+ */
 RZ_API RzBinSection *rz_bin_get_section_at(RzBinObject *o, ut64 off, int va) {
 	RzBinSection *section;
 	RzListIter *iter;
@@ -667,6 +675,61 @@ RZ_API RzBinSection *rz_bin_get_section_at(RzBinObject *o, ut64 off, int va) {
 		}
 	}
 	return NULL;
+}
+
+/**
+ * \brief Find the first binary map at offset \p off .
+ *
+ * \param o Reference to the \p RzBinObject instance
+ * \param off Address to search
+ * \param va When false the offset \p off is considered a physical address, otherwise a virtual address
+ * \return Pointer to a \p RzBinMap containing the address
+ */
+RZ_API RzBinMap *rz_bin_object_get_map_at(RzBinObject *o, ut64 off, bool va) {
+	rz_return_val_if_fail(o, NULL);
+
+	RzBinMap *map;
+	RzListIter *iter;
+	ut64 from, to;
+
+	rz_list_foreach (o->maps, iter, map) {
+		from = va ? rz_bin_object_addr_with_base(o, map->vaddr) : map->paddr;
+		to = from + (va ? map->vsize : map->psize);
+		if (off >= from && off < to) {
+			return map;
+		}
+	}
+	return NULL;
+}
+
+/**
+ * \brief Find all binary maps at offset \p off .
+ *
+ * \param o Reference to the \p RzBinObject instance
+ * \param off Address to search
+ * \param va When false the offset \p off is considered a physical address, otherwise a virtual address
+ * \return Vector of \p RzBinMap pointers
+ */
+RZ_API RzPVector *rz_bin_object_get_maps_at(RzBinObject *o, ut64 off, bool va) {
+	rz_return_val_if_fail(o, NULL);
+
+	RzBinMap *map;
+	RzListIter *iter;
+	ut64 from, to;
+
+	RzPVector *res = rz_pvector_new(NULL);
+	if (!res) {
+		return NULL;
+	}
+
+	rz_list_foreach (o->maps, iter, map) {
+		from = va ? rz_bin_object_addr_with_base(o, map->vaddr) : map->paddr;
+		to = from + (va ? map->vsize : map->psize);
+		if (off >= from && off < to) {
+			rz_pvector_push(res, map);
+		}
+	}
+	return res;
 }
 
 RZ_API RzList *rz_bin_reset_strings(RzBin *bin) {

--- a/librz/bin/bobj.c
+++ b/librz/bin/bobj.c
@@ -961,3 +961,76 @@ err:
 	rz_list_free(sections);
 	return res;
 }
+
+static ut64 map_p2v(RzBinMap *m, ut64 paddr) {
+	ut64 delta = paddr - m->paddr;
+	if (delta >= m->vsize) {
+		return UT64_MAX;
+	}
+	return m->vaddr + delta;
+}
+
+/**
+ * \brief Convert offset in the file to virtual address according to binary mappings
+ *
+ * \param obj Reference to \p RzBinObject
+ * \param paddr Offset in the file
+ * \return Converted offset to virtual address or UT64_MAX if the conversion cannot be done
+ */
+RZ_API ut64 rz_bin_object_p2v(RzBinObject *obj, ut64 paddr) {
+	rz_return_val_if_fail(obj, UT64_MAX);
+	RzBinMap *m = rz_bin_object_get_map_at(obj, paddr, false);
+	if (!m) {
+		return UT64_MAX;
+	}
+
+	return map_p2v(m, paddr);
+}
+
+/**
+ * \brief Convert offset in the file to all possible virtual addresses according to binary mappings
+ *
+ * \param obj Reference to \p RzBinObject
+ * \param paddr Offset in the file
+ * \return Vector containing \p ut64 values of all possible virtual addresses
+ */
+RZ_API RzVector *rz_bin_object_p2v_all(RzBinObject *obj, ut64 paddr) {
+	rz_return_val_if_fail(obj, NULL);
+	RzPVector *maps = rz_bin_object_get_maps_at(obj, paddr, false);
+	if (!maps) {
+		return NULL;
+	}
+
+	RzVector *res = rz_vector_new(sizeof(ut64), NULL, NULL);
+	void **it;
+	rz_pvector_foreach (maps, it) {
+		RzBinMap *map = *(RzBinMap **)it;
+		ut64 vaddr = map_p2v(map, paddr);
+		if (vaddr != UT64_MAX) {
+			rz_vector_push(res, &vaddr);
+		}
+	}
+
+	return res;
+}
+
+/**
+ * \brief Convert virtual address to offset in the file according to binary mappings
+ *
+ * \param obj Reference to \p RzBinObject
+ * \param paddr Virtual address
+ * \return Converted virtual address to offset in the file or UT64_MAX if the conversion cannot be done
+ */
+RZ_API ut64 rz_bin_object_v2p(RzBinObject *obj, ut64 vaddr) {
+	rz_return_val_if_fail(obj, UT64_MAX);
+	RzBinMap *m = rz_bin_object_get_map_at(obj, vaddr, true);
+	if (!m) {
+		return UT64_MAX;
+	}
+
+	ut64 delta = vaddr - m->vaddr;
+	if (delta >= m->psize) {
+		return UT64_MAX;
+	}
+	return m->paddr + delta;
+}

--- a/librz/bin/golang.c
+++ b/librz/bin/golang.c
@@ -124,26 +124,9 @@ static ut64 go_string(ut8 *buffer, size_t size, char **output) {
 	return n + read;
 }
 
-// the maps are not set yet on rz_io, so we cannot use it.
-// therefore virt to phys needs to be done manually
 static st64 io_read_va_at(RzBinFile *bf, ut64 vaddr, ut8 *buffer, ut64 size) {
-	RzBinSection *section = NULL;
-	RzListIter *it = NULL;
-	ut64 paddr = 0;
-	RzList *sections = rz_bin_object_get_sections(bf->o);
-	rz_list_foreach (sections, it, section) {
-		if (!strstr(section->name, "data")) {
-			// ignore any non-data section
-			continue;
-		}
-		ut64 end = section->vaddr + section->vsize;
-		if (vaddr >= section->vaddr && vaddr < end) {
-			paddr = vaddr - section->vaddr;
-			paddr += section->paddr;
-			break;
-		}
-	}
-	if (!paddr) {
+	ut64 paddr = rz_bin_object_v2p(bf->o, vaddr);
+	if (paddr == UT64_MAX) {
 		return -1;
 	}
 	return rz_buf_read_at(bf->buf, paddr, buffer, size);

--- a/librz/bin/p/bin_dmp64.c
+++ b/librz/bin/p/bin_dmp64.c
@@ -80,6 +80,10 @@ static void header(RzBinFile *bf) {
 	}
 }
 
+static RzList *strings(RzBinFile *bf) {
+	return rz_bin_file_strings(bf, 4, false);
+}
+
 static RzList *fields(RzBinFile *bf) {
 	RzList *fields = rz_list_newf((RzListFree)rz_bin_field_free);
 	struct rz_bin_dmp64_obj_t *obj = (struct rz_bin_dmp64_obj_t *)bf->o->bin_obj;
@@ -276,6 +280,7 @@ RzBinPlugin rz_bin_plugin_dmp64 = {
 	.destroy = &destroy,
 	.get_sdb = &get_sdb,
 	.header = &header,
+	.strings = &strings,
 	.info = &info,
 	.load_buffer = &load_buffer,
 	.check_buffer = &check_buffer,

--- a/librz/core/cmd/cmd_debug.c
+++ b/librz/core/cmd/cmd_debug.c
@@ -3246,7 +3246,7 @@ RZ_IPI RzCmdStatus rz_cmd_debug_display_bt_handler(RzCore *core, int argc, const
 			RzAnalysisFunction *fcn = rz_analysis_get_fcn_in(core->analysis, frame->addr, 0);
 			rz_cons_printf("%d  %s sp: %s  %-5d"
 				       "[%s]  %s %s\n",
-				i++, pcstr, spstr, (int)frame->size, fcn ? fcn->name : "??", flagdesc ? flagdesc : "", flagdesc2 ? flagdesc2 : "");
+				i++, pcstr, spstr, (int)frame->size, fcn ? fcn->name : "??", rz_str_get(flagdesc), rz_str_get(flagdesc2));
 			free(flagdesc);
 			free(flagdesc2);
 			free(pcstr);

--- a/librz/include/rz_bin.h
+++ b/librz/include/rz_bin.h
@@ -960,6 +960,11 @@ RZ_API bool rz_bin_object_is_string(RZ_NONNULL RzBinObject *obj, ut64 va);
 RZ_API bool rz_bin_object_is_big_endian(RZ_NONNULL RzBinObject *obj);
 RZ_API bool rz_bin_object_is_static(RZ_NONNULL RzBinObject *obj);
 RZ_API RZ_OWN RzVector *rz_bin_object_sections_mapping_list(RZ_NONNULL RzBinObject *obj);
+RZ_API RZ_OWN RzPVector *rz_bin_object_get_maps_at(RzBinObject *o, ut64 off, bool va);
+RZ_API RzBinMap *rz_bin_object_get_map_at(RZ_NONNULL RzBinObject *o, ut64 off, bool va);
+RZ_API ut64 rz_bin_object_p2v(RZ_NONNULL RzBinObject *obj, ut64 paddr);
+RZ_API RzVector *rz_bin_object_p2v_all(RZ_NONNULL RzBinObject *obj, ut64 paddr);
+RZ_API ut64 rz_bin_object_v2p(RZ_NONNULL RzBinObject *obj, ut64 vaddr);
 
 RZ_API RzBinLanguage rz_bin_language_detect(RzBinFile *binfile);
 RZ_API RzBinLanguage rz_bin_language_to_id(const char *language);
@@ -983,8 +988,6 @@ RZ_API bool rz_bin_file_object_new_from_xtr_data(RzBin *bin, RzBinFile *bf, RzBi
 RZ_API RzBinFile *rz_bin_file_at(RzBin *bin, ut64 addr);
 RZ_API RzBinFile *rz_bin_file_find_by_object_id(RzBin *bin, ut32 binobj_id);
 RZ_API RzList *rz_bin_file_get_symbols(RzBinFile *bf);
-//
-RZ_API ut64 rz_bin_file_get_vaddr(RzBinFile *bf, ut64 paddr, ut64 vaddr);
 // RzBinFile.add
 RZ_API RzBinClass *rz_bin_file_add_class(RzBinFile *binfile, const char *name, const char *super, int view);
 RZ_API RzBinSymbol *rz_bin_file_add_method(RzBinFile *bf, const char *classname, const char *name, int nargs);

--- a/test/integration/test_bin.c
+++ b/test/integration/test_bin.c
@@ -299,12 +299,46 @@ bool test_rz_bin_sections_mapping(void) {
 	mu_end;
 }
 
+bool test_rz_bin_p2v2p(void) {
+	RzBin *bin = rz_bin_new();
+	RzIO *io = rz_io_new();
+	rz_io_bind(io, &bin->iob);
+
+	RzBinOptions opt = { 0 };
+	rz_bin_options_init(&opt, 0, 0, 0, false);
+	RzBinFile *bf = rz_bin_open(bin, "bins/elf/ls", &opt);
+	mu_assert_notnull(bf, "ls binary could not be opened");
+
+	RzBinObject *o = bf->o;
+	mu_assert_eq(rz_bin_object_p2v(o, 0x0001c0f8), 0x0001c0f8, "xstrtoumax string p2v");
+	mu_assert_eq(rz_bin_object_v2p(o, 0x0001c0f8), 0x0001c0f8, "xstrtoumax string v2p");
+	mu_assert_eq(rz_bin_object_p2v(o, 0x00021260), 0x00022260, "obstack_alloc_failed_handler symbol p2v");
+	mu_assert_eq(rz_bin_object_v2p(o, 0x00022260), 0x00021260, "obstack_alloc_failed_handler symbol v2p");
+
+	mu_assert_eq(rz_bin_object_v2p(o, 0xcafebabe), UT64_MAX, "non existing vaddr");
+	mu_assert_eq(rz_bin_object_p2v(o, 0xcafebabe), UT64_MAX, "non existing paddr");
+
+	RzVector *v = rz_bin_object_p2v_all(o, 0xcafebabe);
+	mu_assert_eq(rz_vector_len(v), 0, "non existing paddr should have 0 elements in vector");
+	rz_vector_free(v);
+
+	v = rz_bin_object_p2v_all(o, 0x21260);
+	mu_assert_eq(rz_vector_len(v), 1, "p2v_all of obstack_alloc_failed_handler paddr should have 1 element");
+	mu_assert_eq(*(ut64 *)rz_vector_head(v), 0x22260, "obstack_alloc_failed_handler symbol p2v_all");
+	rz_vector_free(v);
+
+	rz_bin_free(bin);
+	rz_io_free(io);
+	mu_end;
+}
+
 bool all_tests() {
 	mu_run_test(test_rz_bin);
 	mu_run_test(test_rz_bin_reloc_storage);
 	mu_run_test(test_rz_bin_file_delete);
 	mu_run_test(test_rz_bin_file_delete_all);
 	mu_run_test(test_rz_bin_sections_mapping);
+	mu_run_test(test_rz_bin_p2v2p);
 	return tests_passed != tests_run;
 }
 

